
<h1 align ="center">130-30 Portfolio Construction using Metaheuristics</h1>

![](POModels_130_30Strategy_GoalPost.png) 


# 1. What is a 130-30  portfolio?

*130-30 strategy* is a portfolio construction method which ensures investment exposure and market protection at the same time. The strategy adopts *leveraging* by **shorting poor performing stocks to the tune of 30% of the portfolio value** and diverting the funds **to invest in the long,  on better performing stocks to the tune of 130% of the portfolio value**. 
130-30 portfolios are *long-short portfolios* and by and large have been observed to perform better than *long-only portfolios*.

# 2. 130-30 Portfolio Construction

Let us consider an investor who desires to construct a 130-30 portfolio, optimal with regard to an objective function and subject to constraints that reflect restrictions on asset allocation and the investor's preferences. 

    (1)  The objective is to maximize Sharpe Ratio of the portfolio. 

(See  https://github.com/PaiViji/PythonFinance-PortfolioOptimization/blob/master/Lesson6_SharpeRatioOptimization/Lesson6_MainContent.ipynb  to know about Sharpe Ratio based portfolio optimization)

The following **constraints** are imposed on the portfolio by the investor:

    (2) A budget constraint on the portfolio, where the capital is fully invested, (i.e.) the sum of the portfolio     
        weights equals 1.
    (3) A budget constraint of 130% on the capital invested in the long positions, which automatically implies a 
        budget constraint of 30% on the capital invested in the short positions of the  portfolio.
    (4) Leveraged bounds on each long position to lie between [0, 1.3]
    (5) Bounds on short positions to lie between [-0.3,0]
    (6) Portfolio beta needs to be 1 to ensure that the volatility of the portfolio matches with that of the 
        market.
    

# 3. Mathematical Formulation of the Investor's 130-30 Portfolio Model

Let us suppose that the investor desires to construct a portfolio P comprising assets $A_1, A_2, ...A_N$, with $\bar \mu_{(1 X N)} = [\mu_1, \mu_2, ...\mu_N]$ as the asset returns and $\bar W_{(N X1)}=[W_1, W_2, ...W_N]'$ as the weights.

Let $R_f$ be the risk free rate of return.

The portfolio return $r$ determined by a weighted summation of its individual asset returns is given by $\bar \mu . \bar W = \sum\left({\mu_i.W_i}\right)$ and the risk is given by $\sqrt{{\bar W}'.V.\bar W}$, where V is the  variance-covariance matrix of returns.

With regard to portfolios with equity stocks, *portfolio beta* is the weighted sum of the asset betas,where the weights are represented by the portfolio weights. Portfolio beta is given by, $\sum{\beta_i.W_i}$, where $\beta_i$ are the asset betas and $W_i$ are the portfolio weights. 

(See  https://github.com/PaiViji/PythonFinance-PortfolioOptimization/blob/master/Lesson1_FundaRiskReturnPortfolio/Lesson1_MainContent.ipynb, to know about risk, return and asset betas of a portfolio)



The mathematical model for the 130-30 strategy based Portfolio Optimization whose objective function and constraints were defined in the previous section is as shown below:

1.Objective function for maximizing Sharpe Ratio of the portfolio

![](POModels_130_30Strategy_Eqn1_1.png)


subject to the constraints that describe constraints (2)-(6) listed in Sec. 2. viz.,

![](POModels_130_30Strategy_Eqn1_2to1_6.png)

In the above system of equations,  $W^+$ and $W^-$ indicate the weights of the long and short positions of the optimal portfolio. Equations [1.3] to [1.5] illustrate the budget and leveraged bound constraints imposed on the long and short positions. Equation [1.6] denotes the portfolio beta constraint and equation [1.2] describes a fully invested long-short portfolio.

## 3.1 Long-Plus-Short Portfolio vs Integrated Long-Short Portfolio

The 130-30 portfolio model discussed here, is not a mere combination of a *long-only portfolio* with a *short-only portfolio*, where the investor earmarks some positions in the portfolio to be longed and the rest to be shorted. Jacobs et al., [JAC 99] were critical of this approach and while terming it as a
*long-plus-short* portfolio,  asserted that such a portfolio has lesser merits when compared to a real *long-short portfolio* that undertakes an *integrated combination of long and short positions*. They claimed that a long-short portfolio construction is not a two-portfolio strategy but *"...a one-portfolio strategy in which the long and short positions are determined jointly within an optimization that takes into account the expected returns of the individual securities, the standard deviation of those returns  and the correlations between them, as well as the investor's tolerance for risk"* [JAC 99]. 

The strategy adopted to construct such a long-short 130-30 portfolio is evolved using **metaheuristics**. The metaheuristic strategy undertakes an **integrated optimization** of the assets in the portfolio to determine the long and short positions of the portfolio that yield maximal Sharpe Ratio,  subject to all the inherent constraints laid down by the investment strategy as well as the investor. 



# 4.  Metaheuristic solution strategy

We proceed to undertake the integrated optimization of the 130-30 portfolio model described by equations [1.1] to [1.6] employing a robust **metaheuristic strategy** viz., **Differential Evolution with Hall of Fame (DE HOF)**. DE HOF dynamically arrives at the optimal composition of long and short positions of the portfolio that yield the maximal Sharpe Ratio, subject to all the constraints enforced over the assets and the portfolio. 

(See Chapter 2: A Brief Primer on Metaheuristics, of [Pai 2018], to know more about metaheuristics and Differential Evolution)

Metaheuristic strategies build *feasible solution sets* that satisfy the constraints imposed,  during the course of their execution process. However *constraint handling* has been a major problem in the application of metaheuristic strategies to complex constrained optimization problems, to tackle which several methods such as *Penalty Function Strategy* and *Repair Strategy* have been proposed. 

For the 130-30 portfolio optimization, DE HOF makes use of Joines and Houck's *Dynamic Penalty Function Strategy* [JOI 1994] to handle the portfolio beta constraint (equation [1.6]) and Repair Strategies to handle the budget constraints on the portfolio and long positions as well as the leveraged bounds constraints on the long and short positions (equations [1.2] to [1.5]).

Repair strategies are custom made to suit the requirements of the problem and evolving such a strategy that will help satisfy one or more constraints at one go, can turn out to be difficult. Nevertheless, once the strategy is evolved, it can help churn out populations of feasible solution sets leading to faster convergence of the metaheuristic strategy. 

(See Sec. 6.4.2 of Chapter 6: Metaheuristic 130-30 Portfolio Construction [Pai 2018], to know about the Repair Strategies evolved for the 130-30 portfolio optimization model)



# 5.  Transformation of the Mathematical Model 

Joines and Houck's dynamic penalty function strategy is used to tackle the portfolio beta constraint represented by equation [1.6]. The portfolio beta constraint is accomodated in the "penalized" objective function by defining appropriate penalty functions. The transformed mathematical model is shown below. 

![](POModels_130_30Strategy_Eqn1_7.png)

where $ \psi \left ( \bar{W}, {\bar{\beta}_A}, t \right )$ is the constraint violation function that tackles the portfolio beta constraint using dynamic penalty functions and is defined as shown below.

![](POModels_130_30Strategy_Eqn1_8.png)

In the system of equations [1.8], $(C, \alpha, \beta)$ are all constants and the penalty term $(C.t)^\alpha$ increases constantly with each generation count *t* of the metaheuristic strategy DE HOF. $\varphi\left ( \bar{W}, \bar{\beta_A}\right )$ tackles the portfolio beta constraint as $ \left | \sum_{k=1}^{N} \left ( \beta _A^k.W_k \right )-1\right | \leq \epsilon$, for a tolerance limit of $\epsilon$.

DE HOF now strives to solve the optimization model whose penalized objective function is described by equations [1.7]-[1.8] subject to the budget and leveraged bounds constraints imposed on the long and short positions of the portfolio, described by equations [1.2]-[1.5] only. 

To reiterate, DE HOF employs repair strategies to tackle  these constraints resulting in faster convergence. 

# 6. Differential Evolution with Hall of fame - a brief roundup

DE HOF is a population based metaheuristic strategy that can efficiently solve complex constrained optimization problems. We restrict the discussion to the DE HOF designed to solve the 130-30 portfolio optimization problem model. The input, process and output of the DE HOF strategy are as follows:

### Inputs 

It is essential that the *portfolio parameters* and the *DE HOF strategy parameters* are clearly set before the optimization process begins. 

The portfolio parameters are (1) assets in the portfolio (2) asset betas (3) mean asset returns (4) covariance of returns (the variance-covariance matrix of returns) and (5) risk free rate of return.

The DE HOF parameters are (1) population size (2) number of generations (3) dynamic penalty function parameters $(C, \alpha, \beta)$ (4) Scale factor $\beta_S$  and (5) probability of recombination $p_r$.

### Process

DE HOF begins its execution by generating an *initial random population of individuals* that represent the candidate solution sets to the optimization problem concerned. Each individual in the population represents a collection of N *genes* that represent random weights generated in the range [-0.3, 1.3] for the N assets in the portfolio. 

The candidate solution sets are transformed into feasible solution sets by calling the weight repair strategies which repair each individual in the population, so that it satisfies constraints represented by equations [1.2] to [1.5]. At the end of the process the initial population is transformed into a population that is a feasible solution set and is termed as the *parent population* 

Typical of metaheuristic strategies, DE HOF now computes the *fitness function values* of each individual in the parent population making use of the penalized objective function described by equations [1.7]-[1.8].

Making use of mutation and crossover operators ( *Rand4/Best/Dir5 mutation operator* and *Binomial Crossover operator*) and the scale factor $\beta_A$, DE HOF generates what are called *trial vectors* which eventually leads to the production of  the *offspring population* of individuals. (See [FEO 2006] to know more about Rand4/Best/Dir5 operator)

The offspring population of individuals is standardized to satisfy the respective constraints, by calling the weight repair strategies once again and their fitness function values are computed as was done for the parent population.

Based on the fitness function values of the individuals in  the parent population and offspring population, the best fit individuals are selected using the *Deterministic Selection operator* and pushed to the next generation. The best among the better individuals selected for the next generation with maximal fitness function value and zero penalty function value (no constraints violated) is inducted into the *Hall of Fame*. 

At the end of the first generation, the next generation pool of individuals get set as the parent population and the second generation begins. The second generation offspring population is now generated by repeating the reproduction process and the best among the better individuals selected for the third generation competes with the individual in the Hall of Fame. DE HOF ensures that only the best individual by way of maximal fitness function value and zero penalty function value generated this far, is inducted into the Hall of Fame. 

The generation cycles progress  until the termination criterion is met with, at which stage the individual in the Hall of Fame is declared to be the optimal solution to the problem concerned. Since it is integrated optimization,  those positions that are to be longed or shorted to arrive at the optimal 130-30 portfolio,  is known only at this stage.

### Output

The genes of the HOF individual represent the optimal weights or proportion of capital to be invested in the assets of the portfolio.  Positive weights indicate assets that are to be longed and negative weights indicate assets that are to be shorted. Given the mean returns and the covariance of returns, the risk and return of the optimal 130-30 portfolio can be easily computed. It can be verified that all the long and short positions in the optimal portfolio satisfy their respective budgets and leveraged bounds constraints and the portfolio beta equals 1, ensuring that the volatility of the optimal 130-30 portfolio matches that of  the market.

(See Sec. 6.4.3 of Chapter 6 Metaheuristic 130-30 Portfolio Construction [PAI 2018] to know more about the design, process flow chart and  execution of Differential Evolution with Hall of Fame based 130-30 Portfolio construction)


# 7. Case Study

Let us demonstrate the construction of 130-30 long-short portfolios over S&P BSE200 (April 2009- July 2020) data set of Bombay Stock Exchange, India, which includes Covid 19 crisis period as well.   To keep the story short, we assume that the investor has already made a *technically diverse choice of assets in the portfolio* (a **k-portfolio**, in fact) and is ready with the mean returns $\mu$, the variance-covariance matrix of returns V and the  betas of the assets constituting the portfolio.

Note that a *k-portfolio* is an outcome of a **heuristic portfolio selection strategy**,  where the universe of stocks is grouped into clusters that display *intra-class similarity* and *inter-class dissimilarity* with regard to the mean-returns and covariance of returns.  Since assets belonging to a cluster are similar in behavior, the investor now makes a choice of one asset each from each cluster to ensure **diversification of assets in the portfolio**. A clustering technique such as **k-means algorithm** can be used to group the stock universe into k clusters with the investor exercising the choice of k. 

(See  https://github.com/PaiViji/PythonFinance-PortfolioOptimization/blob/master/Lesson3_HeuristicPortfolioSelection/Lesson3_MainContent.ipynb and Chapter 3 Heuristic Portfolio Selection  of [PAI 2018],  to know more about the construction  of k-portfolios and their merits)

Let the k-portfolio selected by the investor comprise the following  20 assets,  after making a heuristic portfolio selection for  k = 20:

Ajanta Pharma Ltd. ["'AJANTPHARM'"], Amara Raja Batteries Ltd.["'AMARAJABAT'"],  Ashok Leyland Ltd.["'ASHOKLEY'"], Bharat Forge Ltd. ["'BHARATFORG'"],  Cipla Ltd. ["'CIPLA'"],  GAIL (India) Ltd. ["'GAIL'"], HDFC Bank Ltd.["'HDFCBANK'"], Hindustan Petroleum Corporation Ltd. [ "'HINDPETRO'"], Vodafone Idea Ltd. ["'IDEA'"], Indraprastha Gas Ltd. ["'IGL'"],  Indian Oil Corporation Ltd.[ "'IOC'"], ITC Ltd. ["'ITC'"],                        JSW Steel Ltd.[ "'JSWSTEEL'"], Motilal Oswal Financial Services Ltd. ["'MOTILALOFS'"], Petronet LNG Ltd. ["'PETRONET'"], State Bank of India Ltd. ["'SBIN'"], Shree Cement Ltd. ["'SHREECEM'"],  Tata Steel Ltd. ["'TATASTEEL'"], Tata Consultancy Services Ltd. ["'TCS'"], Wipro Ltd. ["'WIPRO'"]

The objective is to construct a 130-30 portfolio that will yield maximal Sharpe Ratio subject to the constraints listed in equations [1.2]-[1.6]. 


A fragment of the CSV file  **S&PBSE200_kPortfolio.csv**,  which describes the asset labels, mean returns, variance-covariance matrix of returns and asset betas, to be used by DE HOF for the construction of optimal 130-30 portfolio is shown below:


![](POModels_130_30Strategy_Figure1_1.png)
<h4 align ="center">Fig.1.1 Structure of the input file to DE HOF that captures details about assets in the portfolio, their mean and covariance of returns (daily %) and asset betas</h4>


# 8. Python coding of DE HOF for 130-30 Portfolio Construction

The DE HOF program is a conglormeration of functions typical of any metaheuristic strategy. The functions are listed first followed by the main program, with a brief description of the task accomplished by the function code.

### 8.1   Function   WeightStdzn130_30BoundsConstr

This function implements Phase 1 of the  weight repair strategy that DE HOF uses to standardize or repair the weights of a population individual,  so that all the asset weights lie between [-0.3, 1.3] and their sum equals 1. That is, the budget constraint on the portfolio described by equation [1.2] and the bounds stipulated for a 130-30 portfolio which is [-0.3, 1.3] are met with before Phase 2 of the weight repair strategy, proceeds to repair them further to make them satisfy the constraints described by equations [1.3]-[1.5]. 

The population of individuals each of which represents the weights of the N assets in the portfolio is in reality an array (InpWeightMat) of size (Population Size X N). LowUpBounds is an array of size (2 X N) where  LowUpBounds[0, :] is initialized to -0.3 and LowUpBounds[1,:] is initialized to 1.3.

The function returns the standardized weight matrix StdWeightMatrix, which represents the population of individuals whose weights lie in the interval [-0.3, 1.3] and satisfy the portfolio budget constraint of sum of weights equaling 1 (equation [1.2])

In [2]:
"""
Weight Repair Strategy - Phase 1
-----------------------------------------------------------------------------
Standardization of weights of each individual in the population so that
they satisfy the portfolio budget constraint of
sum of weights equals 1 while each weight lies between [-0.3, 1.3]

Reference: Sec. 6.4.2, Chapter 6 Metaheuristic 130-30 Portfolio Construction of
[PAI, 2018] 
[PAI, 2018] G A Vijayalakshmi Pai, Metaheuristics for Portfolio Optimization-An 
            Introduction using MATLAB, ISTE-Wiley, 2018.
            
MATLAB Version
https://in.mathworks.com/matlabcentral/profile/authors/2806050-dr-g-a-vijayalakshmi-pai
----------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def WeightStdzn130_30BoundsConstr(InpWeightMat, LowUpBounds):
    
    #dependencies
    import numpy as np
    
    WeightMatrix = InpWeightMat
    [RowsMat, ColsMat]=np.shape(WeightMatrix) 
    [LowBound, UpBound] = np.shape(LowUpBounds) 
    
        
    # Steps 1 and 2 of Weight Repair Strategy Phase 1
    # Standardize weights to satisfy their lower bounds while  ensuring
    # that their sum equals 1
    for i in range(RowsMat):
        #R: those weights which are less than their respective lower bounds
        R = [] 
        for j in range(ColsMat):
            
            if (WeightMatrix[i,j]< LowUpBounds[0,j]):
                WeightMatrix[i,j]= LowUpBounds[0,j] 
                R.append(j)
                
        #Q: those weights which satisfy their lower bounds
        Q = list(set(range(ColsMat))-set(R))
        F = 1.0 - np.sum(LowUpBounds[0,R])- np.sum(LowUpBounds[0,Q]) 
        L = np.sum(np.abs(WeightMatrix[i,Q])) 
        if (L==0):
            term = F / len(Q) 
            WeightMatrix[i,Q]= LowUpBounds[0,Q]+ term 
        else:
            term = F / L 
            WeightMatrix[i,Q] = LowUpBounds[0,Q]+ np.abs(WeightMatrix[i,Q])* term 
            
        
    # Steps 3-6 of Weight Repair Strategy Phase 1 
    # standardize upper bounds so that weights ultimately satisfy both upper
    # and lower bounds while their sum equals 1
    for i in range(RowsMat): 
        
        r = [] 
                    
        ExitFlag =  True 
        q = list(set(range(ColsMat))-set(r))
        
        while (ExitFlag ==  True): 
            ExitFlag =  False  
           
            for j in range(len(q)):
                if ( WeightMatrix[i, q[j]] <= LowUpBounds[1, q[j]] ):
                    continue 
                else:
                    ExitFlag =  True 
                    r.append(q[j]) 
                    
            q = list(set(range(ColsMat))-set(r))
            if ( ExitFlag ==  True):
                L =np.sum(np.abs(WeightMatrix[i,q])) 
                F = 1.0 - ( np.sum(LowUpBounds[0,q])+ np.sum( LowUpBounds[1,r]) ) 
                if (L==0): 
                    term = F 
                    WeightMatrix[i,q[0]]= term 
                else:
                    term = F/L 
                    WeightMatrix[i,q] = LowUpBounds[0,q] + (np.abs(WeightMatrix[i,q])* term) 
                    
                WeightMatrix[i,r]=LowUpBounds[1,r] 
                
    StdWeightMatrix = WeightMatrix
    
    
    return StdWeightMatrix


### 8.2  Function  WeightStdzn130_30BudgetConstr  

This function implements Phase 2 of the Weight Repair Strategy. The output of Phase 1 of the Weight Repair Strategy is fed as input to this function. This population of individuals  is now repaired to satisfy the leveraged bounds and budget constraints on the long and short positions of the portfolio described by equations [1.3]-[1.5]. The output array StdzWeightMatrix of this function represents a feasible solution set to the transformed mathematical model, that satisfies all the constraints described by equations [1.2]-[1.5].

In [3]:
"""
Weight Repair Strategy - Phase 2
-----------------------------------------------------------------------------
Pre-requisite: The input population of individuals to this function 
must have undergone Weight Repair Strategy Phase 1 processing, implemented
as function WeightStdzn130_30BoundsConstr.

Standardization of weights of each individual in the population undertaken so that
they satisfy the budget and leveraged bounds constraints on long positions and 
bounds constraints on short positions, subject to the fully invested 
constraint of sum of weights equals 1.

Reference: Sec. 6.4.2, Chapter 6 Metaheuristic 130-30 Portfolio Construction of
[PAI, 2018] 
[PAI, 2018] G A Vijayalakshmi Pai, Metaheuristics for Portfolio Optimization-An 
            Introduction using MATLAB, ISTE-Wiley, 2018.
            
MATLAB Version
https://in.mathworks.com/matlabcentral/profile/authors/2806050-dr-g-a-vijayalakshmi-pai
----------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def WeightStdzn130_30BudgetConstr(InputWeightMatrix, LongLowUpBounds, ShortLowUpBounds):
    
    #dependencies
    import numpy as np
    
    WeightMatrix = InputWeightMatrix
    
    [RowsMat, ColsMat]= np.shape(WeightMatrix)
    StdzWeightMatrix = np.zeros([RowsMat, ColsMat], dtype = float)
    
    #  budgets for long positions
    eta = 1.3  
    gamma = 0  
    for p in range(RowsMat):
        
        #  identify long and short positions in H, where H[0] denotes
        #  indices of long positions and H[1] denotes indices of short positions
        H = GroupAssets130_30(ColsMat, WeightMatrix[p,:]) 
        
        
        #  Adjust weights representing long positions
        DepositWeights = 0.0  
        h = 0  
        SumWeights = np.sum(WeightMatrix[p,H[h]])

        if ((SumWeights <= eta ) & (SumWeights >= gamma)):
            continue  
        else:                       
            DepositWeights = DepositWeights + (SumWeights-eta)  
            MP = eta  
            WeightMatrix[p,:] = StdznAssetWeights130_30LongPositions(WeightMatrix[p,:], H[h], SumWeights, MP, LongLowUpBounds)  
           
        #  Adjust weights representing short positions
        if (DepositWeights ==0):
            continue  
        else:
            h = 1   
            SumWeights = np.sum(WeightMatrix[p,H[h]])  
            AbsSumWeights = np.sum(np.abs(WeightMatrix[p,H[h]]) )  
            DepositWeights = DepositWeights + SumWeights  
            WeightMatrix[p,:] = StdznAssetWeights130_30ShortPositions(WeightMatrix[p,:], H[h], AbsSumWeights, DepositWeights, ShortLowUpBounds)  
           
    StdzWeightMatrix = WeightMatrix
    return StdzWeightMatrix 

### 8.2.1 Function GroupAssets130_30

Since DE HOF undertakes integrated optimization of long-short portfolio, there is no knowing which assets are long or short in each individual of the population. This sub function groups assets in each individual of the population into long and short classes, so that the respective budgets and leveraged bounds can be applied on the respective classes.  

In [4]:
"""
Group assets into long and short positions, given a chromosome of weights
-------------------------------------------------------------------------
@author: Dr G  Vijayalakshmi Pai
"""

def GroupAssets130_30(PortfolioSize, chromosome):
    
    #dependencies
    import numpy as np
    
    #initialization
    LongPositions = []
    ShortPositions = []
    
    for c in range(PortfolioSize):
        if (chromosome[c] >= 0):
            LongPositions.append(c)
        else:
            ShortPositions.append(c)
            
    GroupAssets = [LongPositions, ShortPositions]
     
    return GroupAssets           
         


### 8.2.2   Function  StdznAssetWeights130_30LongPositions 

This sub function takes the weight vector of each individual in the population and repairs them so that the long positions of the corresponding weight vector gathered in M, satisfy the constraints imposed on them viz., the leveraged bounds and budget constraint represented by equations [1.3] and [1.4],  subject to the constraint represented by equation [1.2]. TW denotes the budget imposed on the long positions. 

In [5]:
"""
Weight Repair Strategy - Phase 2
-----------------------------------------------------------------------------
Sub function:
The weights of Long positions of an  individual in the population
are repaired,  to satisfy their respective constraints of leveraged bounds
and budget.

Reference: Sec. 6.4.2, Chapter 6 Metaheuristic 130-30 Portfolio Construction of
[PAI, 2018] 
[PAI, 2018] G A Vijayalakshmi Pai, Metaheuristics for Portfolio Optimization-An 
            Introduction using MATLAB, ISTE-Wiley, 2018.
            
MATLAB Version
https://in.mathworks.com/matlabcentral/profile/authors/2806050-dr-g-a-vijayalakshmi-pai
----------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def  StdznAssetWeights130_30LongPositions(InputWeightVector, M, WeightSum, TW, LowUpBounds):
    
    #dependencies
    import numpy as np
       
    WeightVector = InputWeightVector 
                         
    # adjust lower bounds of assets                        
    L = WeightSum   
    F =  TW - np.sum(LowUpBounds[0,M])
    term = F /L 
    WeightVector[M] = LowUpBounds[0,M]+ (WeightVector[M]* term) 

                     
    # adjust upper bounds of assets 
    r = [] 
    ExitFlag =  True 
    q = M 
    while (ExitFlag ==  True) :
        ExitFlag = False  
        for j in range(len(q)):
            if (WeightVector[q[j]]> LowUpBounds[1,q[j]]):
                ExitFlag =  True 
                r.append(q[j]) 
         
        q = list(set(M)-set(r))
        if (ExitFlag ==  True):
            L = np.sum(WeightVector[q]) 
            F = TW - (np.sum(LowUpBounds[0,q])+ np.sum( LowUpBounds[1, r]) ) 
            term = F/L 
            WeightVector[q] = LowUpBounds[0,q] + (WeightVector[q]* term) 
            WeightVector[r] = LowUpBounds[1,r]

             
    return WeightVector
        

### 8.2.3  Function  StdznAssetWeights130_30ShortPositions

This sub function takes the weight vector whose long position weights have been repaired and proceeds to do a similar activity with regard to the weights of the short positions indicated by M. TW denotes the budget imposed on the short positions.

In [6]:
"""
Weight Repair Strategy - Phase 2
-----------------------------------------------------------------------------
Sub function:
The weights of Short positions of an  individual in the population
are repaired,  to satisfy their respective constraints of leveraged bounds
and budget.

Reference: Sec. 6.4.2, Chapter 6 Metaheuristic 130-30 Portfolio Construction of
[PAI, 2018] 
[PAI, 2018] G A Vijayalakshmi Pai, Metaheuristics for Portfolio Optimization-An 
            Introduction using MATLAB, ISTE-Wiley, 2018.
            
MATLAB Version
https://in.mathworks.com/matlabcentral/profile/authors/2806050-dr-g-a-vijayalakshmi-pai
----------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def StdznAssetWeights130_30ShortPositions(InpWeightVector, M, AbsWeightSum, TW, LowUpBounds):       
      
    #dependencies 
    import numpy as np 
         
    WeightVector = InpWeightVector
      
    # adjust lower bounds of assets, where TW is the budget imposed on
    # short positions
    L = AbsWeightSum   
    F =  TW - np.sum(LowUpBounds[0,M])  
    if (L==0):
        term = F/len(M) 
        WeightVector[M] = LowUpBounds[0,M]+ term 
    else:
        term = F /L 
        WeightVector[M] = LowUpBounds[0, M]+ (np.abs(WeightVector[M])* term)          
      
    # adjust upper bounds of assets 
    r = [] 
    ExitFlag = True 
    q = M
    while (ExitFlag == True):
        ExitFlag = False  
        for j in range(len(q)):
            if (WeightVector[q[j]] > LowUpBounds[1,q[j]]):
                ExitFlag = True 
                r.append (q[j])
         
        q = list(set(M)-set(r))  
        if (ExitFlag == True):
            L = np.sum(np.abs(WeightVector[q])) 
            F = TW - (np.sum(LowUpBounds[0,q])+ np.sum(LowUpBounds[1,r]) ) 
            if  (len(q) <>0):
                if (L==0):
                    term = F 
                    WeightVector[q[0]]= term 
                else:
                    term = F/L 
                    WeightVector[q] = LowUpBounds[0,q] + (np.abs(WeightVector[q])* term) 
            WeightVector[r] = LowUpBounds[1, r] 
         
    return WeightVector
      

### 8.3 Function ConstrViolnFunction130_30

This function computes the constraint violation function defined by equation [1.8] and returns $\varphi\left ( \bar{W}, \bar{\beta_A}\right )$ and G as output (optional). 

In [7]:
"""
Constraint violation function for 130-30 portfolio construction described by 
equation [1.8] in the tranformed mathematical model.

Reference: Chapter 6 Metaheuristic 130-30 Portfolio Construction of [PAI, 2018] 
[PAI, 2018] G A Vijayalakshmi Pai, Metaheuristics for Portfolio Optimization-An 
            Introduction using MATLAB, ISTE-Wiley, 2018.
            
MATLAB Version
https://in.mathworks.com/matlabcentral/profile/authors/2806050-dr-g-a-vijayalakshmi-pai
----------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def  ConstrViolnFunction130_30( WeightMat, BetasAssets, C_param, AlphaParam, beta_param, GenerationCount):
    
    #dependencies
    import numpy as np
    
    [RowMat, ColMat]= np.shape(WeightMat)
    epsilon = 0.001
    
    G = np.zeros(shape =(RowMat))
    psi = np.zeros(shape = (RowMat))
    for i in range(RowMat):  
        
        ChromosomeX = WeightMat[i,:]
        
        # compute penalty function G
        PortfolioBetaTerm = np.sum(np.multiply(BetasAssets, ChromosomeX)) 
        g1Term = np.abs((PortfolioBetaTerm) - 1.0)-epsilon
        
        if (g1Term <=0 ):
            G[i]=0
        else:
            G[i]=1
            
        #compute constraint violation function
        PenaltyTerm= np.power((C_param * GenerationCount), AlphaParam) 
        psi[i] = PenaltyTerm *( G[i] * np.power(g1Term, beta_param))
            
    return [psi, G]
    

### 8.4 Function ComputeFitness130_30

This function computes the fitness function values of the population using the penalized objective function described by equation [1.7].

In [8]:
"""
Compute fitness function values for the population of individuals using the
penalized objective function defined by equation [1.7]

Reference: Chapter 6 Metaheuristic 130-30 Portfolio Construction of [PAI, 2018] 
[PAI, 2018] G A Vijayalakshmi Pai, Metaheuristics for Portfolio Optimization-An 
            Introduction using MATLAB, ISTE-Wiley, 2018.
            
MATLAB Version
https://in.mathworks.com/matlabcentral/profile/authors/2806050-dr-g-a-vijayalakshmi-pai
----------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def ComputeFitness130_30(PoplnMat, ReturnData,  CovarianceData, RiskFreeData,  PsiFunction):
    
    #dependencies
    import numpy as np
    
    [popln_size, cols]= np.shape(PoplnMat)
    weight = np.zeros(cols)
    
    PoplnFitness = np.zeros(popln_size)
    
    for i in range(popln_size):
        weight = PoplnMat[i,:] 
        PoplnFitness[i] = ((np.matmul(ReturnData,  weight.T)-RiskFreeData)/np.sqrt(np.matmul( np.matmul(weight, CovarianceData), weight.T)) ) - PsiFunction[i]  
    
    return PoplnFitness    


### 8.5 Function DEOperatorRand4BestDir5

This function implements the standard Rand4BestDir5 operator of Differential Evolution strategy, which helps generate trial vectors. 

<font color ='green'> (See [FEO 06] to learn more about Rand4BestDir5 operator) </font> 

In [9]:
"""
Differential Evolution's  Rand4BestDir5  operator to generate trial vector population
Reference: Vitaliy Feoktistov, Differential Evolution in Search of Solutions, Springer, 2006.
-----------------------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""
def  DEOperatorRand4BestDir5(popln, FitnessValue,  BetaValue, PoplnSize, IndividualLength):
    
    #dependencies
    import numpy as np
    import array as arr
    
    #initialization
    TrialVectorPopln  = np.zeros([PoplnSize, IndividualLength])
    
    for i in range(PoplnSize):
        DifferentialVecIndex = [0]*5
        V  = np.zeros([4,IndividualLength])
        Vb = np.zeros(IndividualLength)
     
        #set IND the current individual in the population indicated by i
        IND = popln[i,:]
        
     
       # prepare RandomIndex, random number indices for each population
       # individual to enable it choose five random individuals from the
       # population, without repeating itself.
        RandomIndex = arr.array('i', np.random.permutation(PoplnSize))
    
        RandomIndex.remove(i)
        
       # select five random individuals from the population
     
        for u in range (5):
            DifferentialVecIndex[u] = int(RandomIndex[u]) 
         
        
        # Obtain Vb the best individual with the 
        # maximal objective function value and designate the rest as array V
        individuals = [FitnessValue[DifferentialVecIndex[0]], FitnessValue[DifferentialVecIndex[1]], FitnessValue[DifferentialVecIndex[2]], FitnessValue[DifferentialVecIndex[3]], FitnessValue[DifferentialVecIndex[4]]]
        
        max_obj_indx = individuals.index (np.max(individuals))  
        j = 0  
        for z in range(5):
            if (DifferentialVecIndex[z] == DifferentialVecIndex[max_obj_indx] ): 
                Vb = popln[DifferentialVecIndex[z], : ] 
                 
            else:
                V[j,:] = popln[DifferentialVecIndex[z],: ]  
                j=j+1  
        
 
        # obtain trial vector for each of the parent vector individual
        TrialVectorPopln[i,:] = Vb + BetaValue/5 * ( 5*Vb - IND - V[0,:]-V[1,:]-V[2,:]-V[3,:])  
        
    return TrialVectorPopln

### 8.6 Function DEOperatorBinCrossOver

This function implements the standard Binomial Crossover operator of Differential Evolution.

In [10]:
"""
Differential Evolution Binomial Cross over operator
----------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""


def  DEOperatorBinCrossover (ParentPopln, TargetVecPopln, ProbabilityRecombn, Components):
    
    #dependencies
    import numpy as np
    
    [row, col] = np.shape(ParentPopln)
    tau = DEComputeTau (Components, ProbabilityRecombn)
    
    OffspringPopln = np.empty(shape =(row, col), dtype = float)
    for i in range(col):
        if i in tau:
            OffspringPopln[:, i]= TargetVecPopln[:,i]
        else: 
            OffspringPopln[:, i] = ParentPopln[:,i]
            
    return OffspringPopln

### 8.6.1 Function DEComputeTau

This sub function computes a parameter (Tau) of the standard Binomial Crossover operator of Differential Evolution algorithm.

In [11]:
"""
This function computes Tau, a parameter required by Differential Evolution's Binomial Crossover operator
--------------------------------------------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def DEComputeTau(GeneSize, ProbabilityRecombn):
    
    #dependencies
    import numpy as np
    import random
    
    h = list(np.random.permutation(range(GeneSize)))
    
    #set jStar to a random index
    jStar = h[0]   
    
    tau=[jStar] 
    
    h.remove(jStar)
    
    for i in range(GeneSize-1):
        if (random.random()< ProbabilityRecombn):
            tau.append(h[i])
            
    return tau

### 8.7 Function   DEOperatorDetermSelection

This function implements the standard Deterministic Selection operator of Differential Evolution,  which selects the best fit amongst the parent and offspring population and prepares the population of individuals for the next generation.

In [12]:
"""
The standard Deterministic  Selection operator of Differential Evolution
-----------------------------------------------------------------------
@author: Dr G A Vijayalakshmi Pai
"""

def DEOperatorDetermSelection(FeasParentPopln, FeasParentPoplnFitness,PsiParent, FeasOffsprngPopln,FeasOffsprngPoplnFitness, PsiOffsprng,  PoplnSize): 
    
    #dependencies
    import numpy as np
    
    [row, col] = np.shape(FeasParentPopln)
    
    #initialization
    NextGenPool = np.zeros(shape = (PoplnSize, col), dtype = float)
    NextGenPoolFitness = np.zeros(shape = (PoplnSize), dtype = float)
    NextGenPoolPsi = np.zeros(shape = (PoplnSize), dtype = float)
    
    for i in range (PoplnSize):
        if (FeasParentPoplnFitness[i] >= FeasOffsprngPoplnFitness[i]):
            NextGenPool[i,:]= FeasParentPopln[i,:] 
            NextGenPoolPsi[i]= PsiParent[i] 
            NextGenPoolFitness[i] = FeasParentPoplnFitness[i] 
        else: 
            NextGenPool[i,:]= FeasOffsprngPopln [i,:] 
            NextGenPoolPsi[i]=PsiOffsprng[i] 
            NextGenPoolFitness[i] = FeasOffsprngPoplnFitness[i] 
    
    return [NextGenPool, NextGenPoolFitness, NextGenPoolPsi ]                                                               

### 8.8 Main program for DE HOF

The main Python program for DE HOF is shown below. A concise and clear Process Flow Chart of DE HOF constructing the optimal 130-30 portfolio can be found in Chapter 6 Metaheuristic 130-30 Portfolio Construction in [PAI, 2018]. 


In [13]:
"""
                                    Main Program
  130-30 portfolio Construction using Differential Evoution with Hall of Fame Metaheuristic strategy
  --------------------------------------------------------------------------------------------------
     
  Dataset: S&P BSE200 (April 03, 2009 - July 03, 2020)
  
  Input Data File: S&PBSE200_kPortfolio.csv

  The asset labels of the k-portfolio followed by the mean returns of the
  assets, variance-covariance matrix of returns of the assets
  and betas of the assets in the order stated, are available in the 
  input csv file. 
------------------------------------------------------------------     
@author: Dr G A Vijayalakshmi Pai
"""

#dependencies
import numpy as np
import pandas as pd
import array as arr
import csv
import random


# Read csv file to obtain asset labels, mean returns, 
# variance-covariance matrix of asset returns and asset betas
portfolioSize = 20 
rows = 23

stockParamsFileName = 'S&PBSE200_kPortfolio.csv' 
df = pd.read_csv(stockParamsFileName,  nrows= rows)

# extract asset labels
assetLabels = df.columns.tolist()[0:portfolioSize]
print(assetLabels)

# extract mean returns, variance-covariance matrix of returns and asset betas
stockParamsData = np.array(df.iloc[0:, 0:]) 
meanData = np.array(stockParamsData[0,:])
covData = np.array(stockParamsData[1:portfolioSize+1, :])
betaAssets = np.array(stockParamsData[portfolioSize+1,:]) 

# general bounds
bounds = np.vstack( (np.repeat(-0.3, portfolioSize), np.repeat(1.3, portfolioSize)))

# set bounds for the long and short positions in the portfolio as (0, 1.3) and (-0.3, 0) respectively
longPosBounds = np.vstack( (np.repeat(0, portfolioSize), np.repeat(1.3, portfolioSize)))
shortPosBounds = np.vstack( (np.repeat(-0.3, portfolioSize), np.repeat(0, portfolioSize)))

# Risk free rate 
annRiskFree = 6.5/100 
riskfree = (np.power((1+annRiskFree),(1.0/360)) -1.0)*100 
  
# Differential Evolution strategy parameters
poplnSize = 400  
chromosomeLength = portfolioSize  
totalGenerations = 600  
beta = 0.5 
probabilityRecombination = 0.87  

# Joines and Houck's (1994)  dynamic penalty (dp) function strategy
# Set constants C, alpha, beta
CConstant = 0.5  
alphaConstant=2 
betaConstant=2 

# initialize Hall of Fame which will finally hold the optimal weights       
HOFFitness = -999.0 
HOFIndividual = np.zeros(shape=(portfolioSize), dtype = float) 
i1 = 0 
 
# generation counter
generationCount = 1  

# generate initial random population within the range [-0.3, 1.3]
initialPoplnSeed =  np.random.uniform(low = -0.3, high = 1.3, size =(poplnSize, chromosomeLength)) 

# repair weights
initialPoplnBound = WeightStdzn130_30BoundsConstr(initialPoplnSeed, bounds) 
initialPopln = WeightStdzn130_30BudgetConstr(initialPoplnBound, longPosBounds, shortPosBounds) 
# compute constraint violation function
[initialPoplnPsi, initialPoplnG1]= ConstrViolnFunction130_30(initialPopln, betaAssets, CConstant, alphaConstant, betaConstant, generationCount) 
# compute fitness function
initialPoplnFitness = ComputeFitness130_30(initialPopln, meanData, covData, riskfree,  initialPoplnPsi) 
                                      
       
# create parent population and compute fitness       
feasParentPopln = initialPopln  
feasParentPoplnFitness = initialPoplnFitness 
feasParentPoplnPsi = initialPoplnPsi 

#initialize HOF
HOFGenerationArray = np.zeros(poplnSize)
HOFFitnessArray = np.zeros(poplnSize)
# while loop for generation cycles begins      
        
while (generationCount <= totalGenerations):         
        print('generation:',generationCount)                                      
        dynamicBeta = 0.0
        # dynamic beta for each generation
        dynamicBeta = np.random.uniform(low =0.5, high =1.0)
        
        
        # obtain trial vector population
        trialVectorPopln = np.zeros(shape =(poplnSize, chromosomeLength), dtype = float)  
        trialVectorPopln = DEOperatorRand4BestDir5(feasParentPopln, feasParentPoplnFitness,  dynamicBeta, poplnSize, chromosomeLength) 
        
        # obtain offspring population
        offspringPopln = np.zeros(shape =(poplnSize, chromosomeLength), dtype = float)
        offspringPopln = DEOperatorBinCrossover(feasParentPopln, trialVectorPopln, probabilityRecombination, chromosomeLength) 
          
        # undertake weight repair strategy Phase 1
        mutatedPoplnBound = np.zeros(shape =(poplnSize, chromosomeLength), dtype = float)
        mutatedPoplnBound = WeightStdzn130_30BoundsConstr(offspringPopln, bounds) 
        
        # undertake weight repair strategy Phase 2
        feasMutatedPopln = np.zeros(shape =(poplnSize, chromosomeLength), dtype = float)
        feasMutatedPopln = WeightStdzn130_30BudgetConstr(mutatedPoplnBound, longPosBounds, shortPosBounds) 
        
        # compute constraint violation function
        feasMutatedPoplnPsi = np.zeros(shape =(poplnSize), dtype = float)
        feasMutatedPopln_G1 = np.zeros(shape =(poplnSize), dtype = float)
          
        [feasMutatedPoplnPsi, feasMutatedPopln_G1]= ConstrViolnFunction130_30(feasMutatedPopln, betaAssets, CConstant, alphaConstant, betaConstant, generationCount) 
        
          
        # compute fitness function values
        feasMutatedPoplnFitness = np.zeros(shape =(poplnSize, chromosomeLength), dtype = float)
        feasMutatedPoplnFitness = ComputeFitness130_30(feasMutatedPopln, meanData, covData, riskfree,  feasMutatedPoplnPsi) 
        
        # set the population for the next generation
        nextGenPool = np.zeros(shape =(poplnSize, chromosomeLength), dtype = float)
        nextGenPoolFitness = np.zeros(shape =(poplnSize), dtype = float)
        psiFun= np.zeros(shape =(poplnSize), dtype = float)
         
          
        [nextGenPool, nextGenPoolFitness, psiFun ] = DEOperatorDetermSelection(feasParentPopln, feasParentPoplnFitness,  feasParentPoplnPsi,   feasMutatedPopln, feasMutatedPoplnFitness, feasMutatedPoplnPsi,  poplnSize)  
        
        
        # induct best individual into Hall of Fame
        for i in range(poplnSize):
            if (psiFun[i] == 0):
                                       
                    if (nextGenPoolFitness[i] > HOFFitness):
                        HOFFitness = nextGenPoolFitness[i]
                        HOFIndividual = nextGenPool[i,:] 
                        HOFGenerationArray[i1] = generationCount 
                        HOFFitnessArray[i1] = HOFFitness 
                        print('HOF updated Generation', generationCount)
                        print('HOF fitness', HOFFitness)
                        i1=i1+1 
                
        
        # increment generation counter
        generationCount = generationCount + 1
        feasParentPopln = np.zeros(shape =(poplnSize, chromosomeLength), dtype = float)
        feasParentPoplnFitness = np.zeros(shape =(poplnSize), dtype = float)
        
        # set the parent population for the next generation
        feasParentPopln = nextGenPool 
        feasParentPoplnFitness = nextGenPoolFitness 
        
# while loop for DE HOF generations ends
        

# obtain optimal weights from the Hall of Fame 
xStar = HOFIndividual  
    
#   compute optimal portfolio annualized risk and return            
portfolioReturn =  261 * np.sum(np.multiply(meanData, xStar)) 
risk = np.sqrt(261 * np.matmul( np.matmul(xStar, covData), xStar.T)) 
   
print("Integrated 130-30 Portfolio Optimization results:")
print("Annualized risk ( %) ", risk, " Expected Portfolio Annualized return ( %)",  portfolioReturn)

sharpeRatio = (portfolioReturn-annRiskFree*100.0)/ risk
print("Sharpe Ratio", sharpeRatio)

    
longIndex = np.where(xStar > 0)
shortIndex = np.where(xStar < 0)
print("Long positions of the optimal 130-30 portfolio: ", np.array(assetLabels)[longIndex])
print("Short positions of the optimal 130-30 portfolio: ", np.array(assetLabels)[shortIndex])

print("Optimal Weights: long positions: ", xStar[longIndex]) 
print("Optimal Weights: short positions: ", xStar[shortIndex])

print("Sum of long position weights: ",  np.sum(xStar[longIndex]))
print("Sum of short position weights:  ", np.sum(xStar[shortIndex]))
print("Sum of weights: " ,      np.sum(xStar))

portfoliobeta = np.sum(np.multiply(xStar, betaAssets))
print("Portfolio Beta: ", portfoliobeta)

    
print('Successful execution!')



["'AJANTPHARM'", " 'AMARAJABAT'", " 'ASHOKLEY'", " 'BHARATFORG'", " 'CIPLA'", " 'GAIL'", " 'HDFCBANK'", " 'HINDPETRO'", " 'IDEA'", "'IGL'", " 'IOC'", " 'ITC'", " 'JSWSTEEL'", " 'MOTILALOFS'", " 'PETRONET'", " 'SBIN'", " 'SHREECEM'", " 'TATASTEEL'", " 'TCS'", "'WIPRO'"]
('generation:', 1)
('HOF updated Generation', 1)
('HOF fitness', 0.03757187325797837)
('HOF updated Generation', 1)
('HOF fitness', 0.05295606674917502)
('generation:', 2)
('HOF updated Generation', 2)
('HOF fitness', 0.05303038020142848)
('generation:', 3)
('HOF updated Generation', 3)
('HOF fitness', 0.0581358105715034)
('HOF updated Generation', 3)
('HOF fitness', 0.06977451753538741)
('generation:', 4)
('generation:', 5)
('generation:', 6)
('generation:', 7)
('generation:', 8)
('generation:', 9)
('HOF updated Generation', 9)
('HOF fitness', 0.07721780369111188)
('generation:', 10)
('generation:', 11)
('generation:', 12)
('generation:', 13)
('generation:', 14)
('generation:', 15)
('generation:', 16)
('generation:', 17

('generation:', 352)
('generation:', 353)
('generation:', 354)
('generation:', 355)
('generation:', 356)
('generation:', 357)
('generation:', 358)
('generation:', 359)
('generation:', 360)
('generation:', 361)
('generation:', 362)
('generation:', 363)
('generation:', 364)
('generation:', 365)
('generation:', 366)
('generation:', 367)
('generation:', 368)
('generation:', 369)
('generation:', 370)
('generation:', 371)
('generation:', 372)
('generation:', 373)
('generation:', 374)
('generation:', 375)
('generation:', 376)
('generation:', 377)
('generation:', 378)
('generation:', 379)
('generation:', 380)
('generation:', 381)
('generation:', 382)
('generation:', 383)
('generation:', 384)
('generation:', 385)
('generation:', 386)
('generation:', 387)
('generation:', 388)
('generation:', 389)
('generation:', 390)
('generation:', 391)
('generation:', 392)
('generation:', 393)
('generation:', 394)
('generation:', 395)
('generation:', 396)
('generation:', 397)
('generation:', 398)
('generation:

## 9. Optimal 130-30 Portfolio - Analysis of Results

Typical of metaheuristic strategies,  DE HOF yields multiple optimal solutions during its various runs. The interpretation of the results obtained by DE HOF during one of its runs is as follows:


**Optimal 130-30 Portfolio** 

**Maximal Sharpe Ratio:** 1.54     
**Annualized Risk(%):** 26.32     
**Expected Portfolio Annualized Return(%):** 46.90

**Portfolio Beta:** 1.00

**Optimal Weights delivered by DE HOF for the k-portfolio (k = 20):**

(0.225, 0.094, 0.208,  0, -0.063,	-0.237,  0,  0.092, 0, 0.162,  0, 0.055, 0.038, 0, 0.0796,  0, 0.346, 0, 0, 0)


**Long Short Portfolio Composition:**

Fig. 1.2 illustrates the composition of the optimal 130-30 long short portofilio obtained by DE HOF in one of its runs.  The description of the tickers have been listed below.

**Long Positions:**   (Ajanta Pharma Ltd. ["'AJANTPHARM'"] 22.5%, Amara Raja Batteries Ltd.["'AMARAJABAT'"]  9.4%,  Ashok Leyland Ltd.["'ASHOKLEY'"]  20.8%,   Hindustan Petroleum Corporation Ltd. [ "'HINDPETRO'"]  9.2%,  Indraprastha Gas Ltd. ["'IGL'"] 16.2%,   ITC Ltd. ["'ITC'"] 5.5%,  JSW Steel Ltd.[ "'JSWSTEEL'"]  3.8%, Petronet LNG Ltd. ["'PETRONET'"] 7.96%, Shree Cement Ltd. ["'SHREECEM'"]  34.6%.

**Short Positions:**  (Cipla Ltd. ["'CIPLA'"]  6.3%,  GAIL (India) Ltd. ["'GAIL'"]  23.7%)


**Leveraged Budget on long positions:** 130%   **Budget on short positions:** 30%   **Sum of optimal  weights:** 100%


![](POModels_130_30Strategy_Figure1_2(a).png)
<h4 align ="center">(a) Long Positions</h4>

![](POModels_130_30Strategy_Figure1_2(b).png)
<h4 align ="center">(b) Short Positions</h4>
<h4 align ="center">Fig.1.2 130-30 Optimal Long Short Portfolio Composition over S&P BSE200 index</h4>

### 9.1 Multiple solutions - Making a choice!

Table 1.1 shows the summarized characteristics of optimal 130-30 portfolios delivered by DE HOF during various runs. At this juncture it needs to be emphasized that due to the stochastic behavior of exploring solutions over large search spaces by generating a random population of individuals which evolve over the generations into   optimal solutions,  metaheuristic strategies are endowed with inherent capabilities to deliver multiple solutions that are either optimal or acceptable or near-optimal. Thus metaheuristic strategies generate multiple solutions to the problem in hand for various runs. This characteristic of metaheuristic algorithms needs to be viewed meritoriously and  exploited to one's own advantage. 


<h4 align ="center">Table 1.1 Summarized characteristics of optimal 130-30 portfolios of S&PBSE200 index obtained by DE HOF</h4>

![](POModels_130_30Strategy_Table1_1.png)

Thus in the case of 130-30 portfolio construction, though the various runs of DE HOF yield optimal portfolios with similar characteristics of risk, return and Sharpe Ratio, the choice of long short positions and the allocation of weights to the assets in the optimal portolios concerned,  are different and diverse for each run. With the risk, return and Sharpe Ratio of the optimal portfolio guaranteed, it is now left to the discerning investor to make a choice of the portfolio  with a composition of assets and allocation of weights that best suits his or her investment sentiments.

Fig. 1.3 illustrates another optimal 130-30 portfolio constructed by DE HOF during one of its runs. The characteristics of the portfolio are as listed below:


**Another Optimal 130-30 Portfolio** 

**Maximal Sharpe Ratio:** 1.51     
**Annualized Risk(%):** 27.61    
**Expected Portfolio Annualized Return(%):** 48.05

**Portfolio Beta:** 1.00

**Optimal Weights delivered by DE HOF for the k-portfolio (k = 20):**

(0.378, 0.040, 0.086, 0.124, -0.164, -0.136, 0, 0.074, 0, 0.067, 0, 0, 0, 0.142, 0, 0, 0.390, 0, 0, 0)

**Long Short Portfolio Composition:**

**Long Positions:**   Ajanta Pharma Ltd. ["'AJANTPHARM'"] 37.8%, Amara Raja Batteries Ltd.["'AMARAJABAT'"]  4%,  Ashok Leyland Ltd.["'ASHOKLEY'"]  8.6%, Bharat Forge Ltd. ["'BHARATFORG'"] 12.4%,   Hindustan Petroleum Corporation Ltd. [ "'HINDPETRO'"]  7.4%,  Indraprastha Gas Ltd. ["'IGL'"] 6.7%,  Motilal Oswal Financial Services Ltd. ["'MOTILALOFS'"] 14.2%,  Shree Cement Ltd. ["'SHREECEM'"]  39%.

**Short Positions:**  (Cipla Ltd. ["'CIPLA'"]  16.4%,  GAIL (India) Ltd. ["'GAIL'"]  13.6%)


**Leveraged Budget on long positions:** 130%   **Budget on short positions:** 30%   **Sum of optimal  weights:** 100%

![](POModels_130_30Strategy_Figure1_3(a).png)
<h4 align ="center">(a) Long Positions</h4>

![](POModels_130_30Strategy_Figure1_3(b).png)
<h4 align ="center">(b) Short Positions</h4>

<h4 align ="center">Fig.1.3 Another 130-30 Optimal Long Short Portfolio Composition over S&P BSE 200 index</h4>

### 9.2  Convergence characteristics of DE HOF

The convergence behaviour of DE HOF,  where the fitness function values of individuals inducted into the Hall of Fame (that records the best solution obtained this far) during the course of the generation cycles,  for various runs was observed. Fig. 1.4 illustrates the graph and it can be seen that the fitness function values stabilized around 300 generations and thereafter there was no siginificant rise in the fitness function values of individuals inducted into the Hall of Fame. 



![](POModels_130_30Strategy_Figure1_4.png)

<h4 align ="center">Fig. 1.4 Convergence of DE HOF</h4>

## 10.  Conclusion

130-30 portfolios have served as good instruments to ensure investment exposure and market protection at the same time, especially during times when market behaviour has been erratic - the downturns during Covid-19 crisis being a good example in recent times. 

An integrated optimization of long-short portfolios can help  construct optimal 130-30 portfolios with leveraged returns and maximal Sharpe ratios and a metaheuristic strategy such as Differential Evolution with Hall of Fame can serve to construct such a portfolio. 

The metaheuristic strategy is inherently endowed with the potential to deliver multiple near-optimal 130-30 portfolios,  providing investors the luxury of making  choices that follow their investment sentiments. 

## Companion Reading

1. Chapter 6 Metaheuristic 130-30 Portfolio Construction [PAI 2018]


2. Chapter 2 A Brief Primer on Metaheuristics [PAI 2018]


3. MATLAB Demonstration of  130-30 Portfolio Construction using DE HOF,  in Mathworks Central File Exchange https://in.mathworks.com/matlabcentral/fileexchange/64507-metaheuristic-portfolio-optimization-models


4. Sharpe Ratio based Portfolio Optimization 
   https://github.com/PaiViji/PythonFinance-PortfolioOptimization/blob/master/Lesson6_SharpeRatioOptimization/Lesson6_MainContent.ipynb
   
   
5. Heuristic Portfolio Selection
https://github.com/PaiViji/PythonFinance-PortfolioOptimization/blob/master/Lesson3_HeuristicPortfolioSelection/Lesson3_MainContent.ipynb


6. Fundamentals of Risk and Return of a Portfolio
https://github.com/PaiViji/PythonFinance-PortfolioOptimization/blob/master/Lesson1_FundaRiskReturnPortfolio/Lesson1_MainContent.ipynb


## References

1. [FEO 2006]    Feoktistov V, "Differential Evolution in search of solutions", Springer, 2006.


2. [JAC 1999]    Jacobs B I, Levy K N and D Starer, "Long-Short Portfolio Management: An Integrated Approach", The  Journal of Portfolio Management, vol. 25, no. 2, pp. 23-32, 1999.

   
3. [JOI 1994]    Joines J A and C R Houck, "On the use of non-stationary penalty functions to solve nonlinear  constrained optimization problems with GAs", Proceedings of the First IEEE Conference on Evolutionary Computation, pp.579-584, 1994.

   
4. [PAI, 2018]   Vijayalakshmi Pai G. A., "Metaheuristics for Portfolio Optimization- An Introduction using MATLAB", Wiley-ISTE, 2018. https://www.mathworks.com/academia/books/metaheuristics-for-portfolio-optimization-pai.html  
               
               